File: features/f06-composer/validation_test.go
package composer

import (
	"os"
	"slices"
	"strings"
	"testing"
)

func TestValidateCapabilityFixtureCoversEveryContentFamily(t *testing.T) {
	catalog := fixtureCatalog(t)
	imageOne := validImage("image-1", "image/jpeg", 1080, 1080)
	imageTwo := validImage("image-2", "image/jpeg", 1080, 1080)
	video := validVideo("video-1")
	tests := []struct {
		name    string
		content DraftContent
	}{
		{
			name: "text",
			content: DraftContent{
				Text:         "A text post",
				Destinations: []Destination{fixtureDestination("text")},
			},
		},
		{
			name: "link",
			content: DraftContent{
				Text:         "A link post",
				Link:         "https://example.com/post",
				Destinations: []Destination{fixtureDestination("link")},
			},
		},
		{
			name: "image",
			content: DraftContent{
				Text:         "An image",
				Media:        []Media{imageOne},
				Destinations: []Destination{fixtureDestination("image")},
			},
		},
		{
			name: "carousel",
			content: DraftContent{
				Text:         "A carousel",
				Media:        []Media{imageOne, imageTwo},
				Destinations: []Destination{fixtureDestination("carousel")},
			},
		},
		{
			name: "video with capability-specific field",
			content: DraftContent{
				Text:  "A video",
				Media: []Media{video},
				Destinations: []Destination{func() Destination {
					destination := fixtureDestination("video")
					destination.Fields = map[string]string{"visibility": "public"}
					return destination
				}()},
			},
		},
		{
			name: "short video",
			content: DraftContent{
				Text:         "A short",
				Media:        []Media{video},
				Destinations: []Destination{fixtureDestination("short_video")},
			},
		},
		{
			name: "thread",
			content: DraftContent{
				Thread:       []ThreadItem{{Text: "one"}, {Text: "two"}},
				Destinations: []Destination{fixtureDestination("thread")},
			},
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			report := Validate(test.content, catalog)
			if !report.Valid {
				t.Fatalf("valid fixture rejected: %#v", report)
			}
			if report.CapabilityVersion != catalog.Version {
				t.Fatalf("capability version = %q", report.CapabilityVersion)
			}
		})
	}
}

func TestValidateReturnsIndependentDestinationErrorsWithRemedies(t *testing.T) {
	catalog := fixtureCatalog(t)
	content := DraftContent{
		Text:  "Shared",
		Media: []Media{validImage("image", "image/png", 1080, 1080)},
		Destinations: []Destination{
			fixtureDestination("image"),
			fixtureDestination("carousel"),
		},
	}
	report := Validate(content, catalog)
	if report.Valid || !report.Destinations[0].Valid || report.Destinations[1].Valid {
		t.Fatalf("destination outcomes = %#v", report.Destinations)
	}
	error := report.Destinations[1].Errors[0]
	if error.DestinationID == "" || error.Field == "" ||
		error.Rule == "" || error.Remedy == "" {
		t.Fatalf("error is not actionable: %#v", error)
	}
}

func TestValidateFailsClosedForUnknownOrUnavailableCapabilities(t *testing.T) {
	catalog := fixtureCatalog(t)
	catalog.Capabilities = append(catalog.Capabilities, ContentCapability{
		ID: "fixture:blocked", Provider: "fixture",
		ChannelType: "blocked", Format: FormatText,
		Available: false, UnavailableReason: "external review pending",
		Text: TextRules{Allowed: true},
	})
	for name, destination := range map[string]Destination{
		"unknown": {
			ID: "unknown", ChannelID: "channel",
			ChannelType: "unknown", CapabilityID: "missing", Format: FormatText,
		},
		"unavailable": {
			ID: "blocked", ChannelID: "channel",
			ChannelType: "blocked", CapabilityID: "fixture:blocked", Format: FormatText,
		},
	} {
		t.Run(name, func(t *testing.T) {
			report := Validate(DraftContent{
				Text: "content", Destinations: []Destination{destination},
			}, catalog)
			if report.Valid {
				t.Fatal("fail-closed capability accepted")
			}
		})
	}
}

func TestValidateLinkAndCapabilitySpecificFieldRules(t *testing.T) {
	catalog := fixtureCatalog(t)
	linkDestination := fixtureDestination("link")
	report := Validate(DraftContent{
		Link:         "https://192.168.1.2/private",
		Destinations: []Destination{linkDestination},
	}, catalog)
	assertErrorCode(t, report.Destinations[0].Errors, "url_host_not_public")

	videoDestination := fixtureDestination("video")
	videoDestination.Fields = map[string]string{
		"visibility": "friends",
		"undeclared": "value",
	}
	report = Validate(DraftContent{
		Media:        []Media{validVideo("video")},
		Destinations: []Destination{videoDestination},
	}, catalog)
	assertErrorCode(t, report.Destinations[0].Errors, "destination_field_invalid")
	assertErrorCode(t, report.Destinations[0].Errors, "destination_field_unknown")
}

func TestValidateCountsNFCUnicodeCodePoints(t *testing.T) {
	catalog := fixtureCatalog(t)
	destination := fixtureDestination("text")
	report := Validate(DraftContent{
		Text:         strings.Repeat("e\u0301", 280),
		Destinations: []Destination{destination},
	}, catalog)
	if !report.Valid {
		t.Fatalf("NFC text rejected: %#v", report.Destinations[0].Errors)
	}
}

func TestValidateRequiresDestination(t *testing.T) {
	report := Validate(DraftContent{}, fixtureCatalog(t))
	if report.Valid || len(report.Errors) != 1 {
		t.Fatalf("report = %#v", report)
	}
	assertErrorCode(t, report.Errors, "destinations_required")
}

func fixtureCatalog(t *testing.T) CapabilityCatalog {
	t.Helper()
	encoded, err := os.ReadFile("test/fixtures/capabilities.json")
	if err != nil {
		t.Fatal(err)
	}
	catalog, err := ParseCapabilityCatalog(string(encoded))
	if err != nil {
		t.Fatal(err)
	}
	return catalog
}

func fixtureDestination(family string) Destination {
	format := Format(family)
	channelFamily := family
	if family == "short_video" {
		channelFamily = "short"
	}
	return Destination{
		ID:           "destination-" + family,
		ChannelID:    "channel-" + family,
		ChannelType:  ChannelType("fixture_" + channelFamily + "_channel"),
		CapabilityID: "fixture:" + family,
		Format:       format,
	}
}

func assertErrorCode(t *testing.T, errors []ValidationError, code string) {
	t.Helper()
	codes := make([]string, len(errors))
	for index, item := range errors {
		codes[index] = item.Code
	}
	if !slices.Contains(codes, code) {
		t.Fatalf("missing error %q in %v", code, codes)
	}
}

func validImage(id, contentType string, width, height int) Media {
	return Media{
		ID:               id,
		Kind:             MediaImage,
		ContentType:      contentType,
		SizeBytes:        2 * 1024 * 1024,
		Width:            width,
		Height:           height,
		InspectionStatus: InspectionReady,
		URL:              "/api/v1/media/" + id,
	}
}

func validVideo(id string) Media {
	return Media{
		ID:               id,
		Kind:             MediaVideo,
		ContentType:      "video/mp4",
		SizeBytes:        20 * 1024 * 1024,
		Width:            1080,
		Height:           1920,
		VideoCodec:       "h264",
		AudioCodec:       "aac",
		DurationSeconds:  30,
		HasAudio:         true,
		InspectionStatus: InspectionReady,
		URL:              "/api/v1/media/" + id,
	}
}

File: features/f06-composer/test/validation.test.ts
import assert from 'node:assert/strict'
import { readFileSync } from 'node:fs'
import test from 'node:test'

import type {
  CapabilityCatalog,
  ComposerFormat,
  Destination,
  DraftContent,
  Media,
} from '../client/contracts.ts'
import { validateDraft } from '../client/validation.ts'

const catalog = JSON.parse(
  readFileSync(new URL('./fixtures/capabilities.json', import.meta.url), 'utf8'),
) as CapabilityCatalog

test('browser validator covers every provider-agnostic capability family', () => {
  const imageOne = validImage('image-1')
  const imageTwo = validImage('image-2')
  const video = validVideo('video-1')
  const cases: DraftContent[] = [
    content({ text: 'text', destinations: [destination('text')] }),
    content({
      text: 'link',
      link: 'https://example.com/post',
      destinations: [destination('link')],
    }),
    content({
      text: 'image',
      media: [imageOne],
      destinations: [destination('image')],
    }),
    content({
      text: 'carousel',
      media: [imageOne, imageTwo],
      destinations: [destination('carousel')],
    }),
    content({
      text: 'video',
      media: [video],
      destinations: [
        { ...destination('video'), fields: { visibility: 'public' } },
      ],
    }),
    content({
      text: 'short',
      media: [video],
      destinations: [destination('short_video')],
    }),
    content({
      thread: [
        { text: 'one', media_ids: [] },
        { text: 'two', media_ids: [] },
      ],
      destinations: [destination('thread')],
    }),
  ]
  cases.forEach((candidate) => {
    const report = validateDraft(candidate, catalog)
    assert.equal(report.valid, true, JSON.stringify(report))
    assert.equal(report.capability_version, catalog.version)
  })
})

test('browser validation fails closed and returns actionable per-destination errors', () => {
  const report = validateDraft(
    content({
      text: 'content',
      destinations: [
        {
          id: 'unknown',
          channel_id: 'channel',
          channel_type: 'unknown',
          capability_id: 'missing',
          format: 'text',
        },
      ],
    }),
    catalog,
  )
  assert.equal(report.valid, false)
  const error = report.destinations[0]?.errors[0]
  assert.equal(error?.code, 'capability_unknown')
  assert.notEqual(error?.field, '')
  assert.notEqual(error?.rule, '')
  assert.notEqual(error?.remedy, '')
})

test('browser validation enforces safe links and D2-defined custom fields', () => {
  const unsafeLink = validateDraft(
    content({
      link: 'https://192.168.1.20/private',
      destinations: [destination('link')],
    }),
    catalog,
  )
  assert.ok(
    unsafeLink.destinations[0]?.errors.some(
      (error) => error.code === 'url_host_not_public',
    ),
  )

  const fields = validateDraft(
    content({
      media: [validVideo('video')],
      destinations: [
        {
          ...destination('video'),
          fields: { visibility: 'friends', undeclared: 'value' },
        },
      ],
    }),
    catalog,
  )
  const codes = fields.destinations[0]?.errors.map((error) => error.code) ?? []
  assert.ok(codes.includes('destination_field_invalid'))
  assert.ok(codes.includes('destination_field_unknown'))
})

test('browser validation preserves NFC code point parity with the server', () => {
  const report = validateDraft(
    content({
      text: 'e\u0301'.repeat(280),
      destinations: [destination('text')],
    }),
    catalog,
  )
  assert.equal(report.valid, true, JSON.stringify(report))
})

function content(overrides: Partial<DraftContent>): DraftContent {
  return {
    text: '',
    link: '',
    media: [],
    thread: [],
    destinations: [],
    ...overrides,
  }
}

function destination(family: ComposerFormat): Destination {
  const channelFamily = family === 'short_video' ? 'short' : family
  return {
    id: `destination-${family}`,
    channel_id: `channel-${family}`,
    channel_type: `fixture_${channelFamily}_channel`,
    capability_id: `fixture:${family}`,
    format: family,
  }
}

function validImage(id: string): Media {
  return {
    id,
    kind: 'image',
    content_type: 'image/jpeg',
    size_bytes: 2 * 1024 * 1024,
    width: 1080,
    height: 1080,
    inspection_status: 'ready',
    url: `/api/v1/media/${id}`,
  }
}

function validVideo(id: string): Media {
  return {
    id,
    kind: 'video',
    content_type: 'video/mp4',
    size_bytes: 20 * 1024 * 1024,
    width: 1080,
    height: 1920,
    video_codec: 'h264',
    audio_codec: 'aac',
    duration_seconds: 30,
    has_audio: true,
    inspection_status: 'ready',
    url: `/api/v1/media/${id}`,
  }
}
